src/models.py

```diff
@@ -25,6 +28,60 @@ def _verify_bbox_and_pages(self) -> None:
         if self.bounding_boxes is not None and self.pages is not None:
             assert len(self.bounding_boxes) == len(self.pages)
 
+    def merge(self, other: "Chunk", text_separator: str = " ") -> "Chunk":
+        """
+        Merge the chunk with another chunk.
+
+        The ID and chunk type are taken from the first chunk.
+        """
+
+        nullable_properties_with_issue = [
+            property_name
+            for property_name in ["bounding_boxes", "pages"]
+            if (getattr(self, property_name) is None)
+            != (getattr(other, property_name) is None)
+        ]
+
+        optional_properties_with_issue = [
+            property_name
+            for property_name in ["heading", "tokens", "serialized_text"]
+            if (getattr(self, property_name) is None)
+            or (getattr(other, property_name) is None)
+        ]
+
+        if nullable_properties_with_issue:
+            raise ValueError(
+                f"Properties {nullable_properties_with_issue} of chunks being merged must be either both None or both not None."
+            )
+
+        if optional_properties_with_issue:
+            logger.warning(
+                f"Properties {optional_properties_with_issue} of chunks being merged have been set for one or more chunks. These properties will be lost in the merge."
+            )
+
+        combined_bounding_boxes = (
+            self.bounding_boxes + other.bounding_boxes
+            if (self.bounding_boxes is not None) and (other.bounding_boxes is not None)
+            else None
+        )
+        combined_pages = (
+            self.pages + other.pages
+            if (self.pages is not None) and (other.pages is not None)
+            else None
+        )
+
+        return Chunk(
+            # TODO: can we better handle IDs when merging chunks?
+            id=self.id,
```

Comment:
interesting question! will ponder

Reply:
the frontrunner in my head at the moment is something like `tuple[float, float]`. The first float is the index of the passage in reading order, and the second is `0` by default and can be used in case a chunk needs to be split.

For merging two chunks, I'm imagining that the ID of the first chunk could always be taken. (I'm not sure what happens if two chunks that are not adjacent are merged, but I can't think of a reason why that might happen yet!)

For splitting a chunk with ID `(1,0)`, the first chunk would remain as `(1,0)` and the second `(1,1)`.

_Maybe_ a non-string ID is weird but I think I like it - it escapes string-templated identifiers for which you have to write functions to decompose later... and if we really want a string ID we just implement a method to get it